File: examples/continuous_occupancy_quickstart.cpp
#include <rbfsafe/rbfsafe.h>

#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <utility>
#include <vector>

namespace {

template <typename T> T require(rbfsafe::Result<T> result) {
    if (!result)
        throw std::runtime_error(result.error().describe());
    return std::move(result).value();
}

void require(rbfsafe::Result<void> result) {
    if (!result)
        throw std::runtime_error(result.error().describe());
}

} // namespace

int main(int argc, char** argv) {
    using namespace rbfsafe;

    try {
        const std::filesystem::path output = argc > 1
                                                 ? std::filesystem::path(argv[1])
                                                 : std::filesystem::path("continuous-fleet-occupancy.json");
        if (std::filesystem::exists(output)) {
            std::cerr << "output already exists: " << output << '\n';
            return 2;
        }
        const auto robot = require(SerialRobotModel::create(
            "planar-2r",
            {{0.0, 1.0, 0.0, 0.0, JointType::Revolute}, {0.0, 1.0, 0.0, 0.0, JointType::Revolute}},
            {{-1.5, 1.5}, {-1.5, 1.5}}, {0.05, 0.05}));
        const std::vector<TimedConfiguration> trajectory{
            {0, {-0.2, 0.1}},
            {32, {0.2, -0.1}},
        };
        ContinuousOccupancyBuildOptions build_options;
        build_options.maximum_normalized_joint_width = 0.05;
        const auto first =
            require(build_robot_trajectory_occupancy(robot, "fixture-cell-clock-v1", "fixture-cell-world",
                                                     "arm-a", {-4.0, 0.0, 0.0}, trajectory, build_options));
        const auto second =
            require(build_robot_trajectory_occupancy(robot, "fixture-cell-clock-v1", "fixture-cell-world",
                                                     "arm-b", {4.0, 0.0, 0.0}, trajectory, build_options));
        require(verify_robot_trajectory_occupancy(robot, first));
        require(verify_robot_trajectory_occupancy(robot, second));
        ContinuousFleetOccupancyOptions analysis_options;
        analysis_options.minimum_separation = 1.0;
        const auto bundle =
            require(ContinuousFleetOccupancyBundle::create({second, first}, analysis_options));
        require(bundle.save(output));
        std::cout << "bundle=" << bundle.id() << '\n'
                  << "report=" << bundle.report().id << '\n'
                  << "status=" << continuous_fleet_occupancy_status_name(bundle.report().status) << '\n'
                  << "occupancies=" << bundle.occupancies().size() << '\n'
                  << "evidence=unknown\n"
                  << "authorizes_execution=false\n";
        return 0;
    } catch (const std::exception& error) {
        std::cerr << error.what() << '\n';
        return 1;
    }
}

File: examples/continuous_occupancy_quickstart.py
"""Build and persist a conservative two-robot swept-link occupancy bundle."""

from __future__ import annotations

import argparse
from pathlib import Path

import rbfsafe


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "output",
        type=Path,
        nargs="?",
        default=Path("continuous-fleet-occupancy.json"),
    )
    return parser


def main() -> int:
    output = build_parser().parse_args().output
    if output.exists():
        raise FileExistsError(f"output already exists: {output}")
    robot = rbfsafe.SerialRobotModel(
        "planar-2r",
        [
            rbfsafe.DhJoint(
                0.0, 1.0, 0.0, 0.0, rbfsafe.JointType.REVOLUTE
            ),
            rbfsafe.DhJoint(
                0.0, 1.0, 0.0, 0.0, rbfsafe.JointType.REVOLUTE
            ),
        ],
        [rbfsafe.Interval(-1.5, 1.5), rbfsafe.Interval(-1.5, 1.5)],
        [0.05, 0.05],
    )
    trajectory = [
        rbfsafe.TimedConfiguration(0, [-0.2, 0.1]),
        rbfsafe.TimedConfiguration(32, [0.2, -0.1]),
    ]
    build_options = rbfsafe.ContinuousOccupancyBuildOptions()
    build_options.maximum_normalized_joint_width = 0.05
    first = rbfsafe.build_robot_trajectory_occupancy(
        robot,
        "fixture-cell-clock-v1",
        "fixture-cell-world",
        "arm-a",
        [-4.0, 0.0, 0.0],
        trajectory,
        build_options,
    )
    second = rbfsafe.build_robot_trajectory_occupancy(
        robot,
        "fixture-cell-clock-v1",
        "fixture-cell-world",
        "arm-b",
        [4.0, 0.0, 0.0],
        trajectory,
        build_options,
    )
    rbfsafe.verify_robot_trajectory_occupancy(robot, first)
    rbfsafe.verify_robot_trajectory_occupancy(robot, second)
    analysis_options = rbfsafe.ContinuousFleetOccupancyOptions()
    analysis_options.minimum_separation = 1.0
    bundle = rbfsafe.ContinuousFleetOccupancyBundle.create(
        [second, first], analysis_options
    )
    bundle.save(output)
    print(f"bundle={bundle.id}")
    print(f"report={bundle.report.id}")
    print(
        "status="
        + rbfsafe.continuous_fleet_occupancy_status_name(
            bundle.report.status
        )
    )
    print("evidence=unknown")
    print("authorizes_execution=false")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
